ruby/phase-1.md

```diff
@@ -0,0 +1,434 @@
+    *args, # any args to be passed as the input to the workflow
+    options: {}, # execution options overrides
+    **kwargs # keyword arguments as inputs are also supported
+  ) -> WorkflowHandle
+
+  def workflow_handle(namespace, workflow_id, run_id = nil) -> WorkflowHandle
+end
+```
+
+Notes:
+
+- This is not a complete definition of the client
+- Both named and keyword arguments are supported, except for the reserved `options:`
+
+The combination of the singleton client and implicit execution attributes allows us to optionally
+support direct workflow invocations (removing all SDK traces out of the way):
+
+```ruby
+MyWorkflow.start(arg_1, arg_2, keyword_arg: arg_3)
+```
+
+Once a workflow is started all the interactions will be going through the handle object returned to
+the caller. The handle can of course be created from a namespace/workflow_id/run_id combination
+using the `Client#workflow_handle` method. Here the interface of the handle:
+
+```ruby
+class WorkflowHandle
+  def initialize(connection, namespace, workflow_id, run_id = nil)
+
+  def result(await: false, timeout: nil) -> Any
+
+  def describe -> WorkflowExecutionInfo
+
+  def cancel -> void
+
+  def query(name, *args) -> Any
+
+  def signal(name, *args) -> void
+
+  def terminate(reason, details: nil) -> void
+end
+
+```
+
+
+## Worker
+
+The worker interface allows SDK users to register any number of both workflows and activities. These
+will then be polled for on the specified namespaces & task queues and executed. The worker can poll
+on multiple task queues at the same time. These will be determined by the registered activities and
+workflow. If there are any overlaps we'll only create one poller for each unique task queue.
+
+```ruby
+config Temporal::Configuration.new(url: 'localhost:7233')
+worker = Temporal::Worker.new(config)
+
+worker.register_workflow(NewWorkflow)
+worker.register_workflow(NewWorkflow, task_queue: 'tq-3')
+worker.register_activity(AnotherActivity)
+worker.register_activity(AnotherActivity, namespace: 'ns-3')
+
+# Calling Worker#run will block the main thread until receiving SIGTERM or SIGINT
+worker.run
+```
+
+Notes:
+
+- The worker will poll for workflows on `ns-1`/`tq-1` and `ns-1`/`tq-3`
+- And activities on `ns-2`/`tq-2` and `ns-3`/`tq-2`
+- The workflows/activities can of course be defined in any number of different workers
+- The worker is intended to take over the process
+
+### Concurrency model
+
+Ruby can be a tricky language for running concurrent applications. It supports kernel threads,
+however it has a Global Interpreter Lock (GIL) in place which prevents any two threads to be
+executed at the same time (making it effectively single-threaded). The good news is that GIL does
+allow IO-bound instructions to be executed in parallel.
+
+Because of these limitations the most resource efficient solution is a combination of process
+forking and threading. The exact ratio can depend heavily on the application and hardware used. This
```

Comment:
Are you suggesting using OS threads? Those are pretty heavyweight.
Does Ruby have a popular async IO framework that we could leverage? (I'm not familiar with the Ruby ecosystem).

Reply:
I am suggesting OS threads, yet. There are more heavyweight than say goroutines, yes. But since OS threads share memory (unlike processes) these are still an acceptable way of doing concurrency (many languages follow this approach).

Ruby has reactor-based libraries [eventmachine](https://github.com/eventmachine/eventmachine) and [async](https://github.com/socketry/async), however these don't really solve the problem as you still have to wrap developer's code in OS threads since you can't force SDK users to write their activities asynchronously.

However as I mentioned we're using a thread-pool which means that threads are created once and then reused (we are not creating/destroying threads after the initial worker dispatch).

Reply:
This is more relevant to workflows, I would suggest that we try to build the SDK in a way that would allow caching as many workflows as possible.
Using OS threads (vs. coroutines) will hurt the cache capacity.

Reply:
I'm not sure what you are proposing as an alternative. There's really no way to execute code concurrently in Ruby without using Threads (and in Ruby there are only OS threads available). Reactors will force the client code to be written in a very specific way, otherwise it will be blocking the reactor's core

Reply:
What about fibers? Is that an option?

Reply:
Also, are there any ways to isolate workflows to prevent them from sharing global state and accessing non deterministic apis?

Reply:
We will be using Fibers for workflows to implement a manual reactor (stopping execution when workflow is waiting on something). But Fibers are blocking the thread they are called on, which means that you can't execute more than a single Fiber at a time on the same thread. Here's an example:

threads:
```ruby
thread = Thread.new do
  puts 'Thread started'
  sleep 1
  puts 'Thread finished'
end

sleep 0.5
puts 'Waiting on thread'
thread.join # blocking main thread until our `thread` finishes
puts 'All finished'
```
This will print:

```
=> Thread started
=> Waiting on thread
=> Thread finished
=> All finished
```

fibers:
```ruby
fiber = Fiber.new do
  puts 'Fiber started'
  sleep 1
  puts 'Fiber finished'
end

sleep 0.5
puts 'Waiting on fiber'
fiber.resume # blocking main thread to execute a fiber
puts 'All finished'
```

this however will print:

```
=> Waiting on fiber
=> Fiber started
=> Fiber finished
=> All finished
```

Basically the Fiber does not start itself (unlike a thread) until you call `Fiber#resume` which in turn blocks the main thread.